### ISSUE TITLE
Usecase of createView

### ISSUE THREAD

**Issue by erikzenker** [ComputationalRadiationPhysics/alpaka#106]
I was on the way to create the next example for alpaka and was wondering
what the views which are created with `createView` are used for. The only
use case I see is to be able to copy an n-dim buffer into an m-dim buffer ( n!=m). Are
there more usecases ?

**Comment by BenjaminW3** [comment]
The view concept is the base-concept of the buffer concept. The only additional thing a buffer provides is the allocation and deallocation of memory.
The ~~`ViewBasic`~~ `ViewSubView` class that is currently used allows to select a subregion of an other view (possibly a buffer) by providing the new extent and the offset within the baseView.
~~NOTE: This is not tested nor fully designed.
TODO: The `BufPlainPtrWrapper` is no buffer and should be renamed to `ViewPlainPtr` and moved to the views.~~ done

**Comment by erikzenker** [comment]
How can I access the the extent of a view ?

**Comment by erikzenker** [comment]
Furthermore, do you have a idea how we could use the view concept in a useful example ?

**Comment by BenjaminW3** [comment]
You can get the extent of a view/buffer by calling either `extent::getWidth`, `extent::getHeight`, `extent::getDepth` or directly `extent::getExtent<Tidx>`. Additionally there is `getProductOfExtent` which may better be renamed to `getExtentProduct`. To get all extents as a vector there is `extent::getExtentVec`.

The view concept is useful because every type can specialize the view traits and can then be used in all alpaka methods that expect views. E.g. you can call `alpaka::mem::view::copy` with any combination of types that already have these specializations which are `BufCpu`, `BufCudaRt`, `std::vector` `std::array`, `BufPlainPtrWrapper` and `ViewBasic` without wrapping their plane pointers within a `BufPlainPtrWrapper` first.

I will rename and move the file `BufStdContainers.hpp` to `ViewStdContainers.hpp` as well as `BufPlainPtrWrapper.hpp` to `ViewPlainPtr.hpp` in the next days after #107 has been merged to correct the namings.

**Comment by erikzenker** [comment]
I can only guess with which kind of arguments I need to feed the functions `extens::getWidth(arg)` etc.
The documentation tells me that `arg` needs to be of the type `TExtent` which can be a view or a buffer or what ever implements the functionality, right ? What I am currently not able to guess is what is meant by `Tidx`. How do I need to feed that function ?

Can you have a look into the following code? It creates a buffer with three dimensions (2,2,2)
and creates on one dimensional view afterwards.

    const alpaka::Vec<DimMem, Size> extents3D(static_cast<Size>(nElementsPerDim),
                                              static_cast<Size>(nElementsPerDim),
                                              static_cast<Size>(nElementsPerDim));
    
    alpaka::mem::buf::Buf<DevHost, Data, DimMem, Size> hostBuffer (alpaka::mem::buf::alloc<Data, Size>(devHost, extents3D));

    using DataView1D = alpaka::mem::view::View<DevHost,
                                               Data,
                                               alpaka::dim::DimInt<1>,
                                               Size>;

    auto hostBufferView1D = alpaka::mem::view::createView<DataView1D>(hostBuffer);

    std::cout << alpaka::extent::getWidth(hostBufferView1D) << std::endl;
    std::cout << alpaka::extent::getHeight(hostBufferView1D) << std::endl;
    std::cout << alpaka::extent::getDepth(hostBufferView1D) << std::endl;   

The output is:

    2
    1
    1

So create view just virtually deletes the second and third dimension from the buffer ? Or
does the output need to look like: 

    8
    1
    1

**Comment by BenjaminW3** [comment]
Currently (as far as I remember because I have created the ticket #40) the view does not support dimension changes. The wrapping view would have to check that the corresponding pitches do not have any padding while doing so. Therefore, I would say that it is currently a bug that your code is compiling. However, I would not want to disable it now, but rather implement the missing functionality. (I might have a look at it while correcting the buf->view names)

**Comment by bernhardmgruber** [comment]
I think alpaka's `ViewSubView` should be entirely replaced by `std::mdspan`, which would then give us the features discussed in this issue.

**Comment by j-stephan** [comment]
> I think alpaka's `ViewSubView` should be entirely replaced by `std::mdspan`

No. This would break user code that cannot depend on `std::mdspan`.

Edit: At least not until `std::mdspan` becomes available in the standard libraries.

**Comment by fwyzard** [comment]
I'd also like to keep `mdspan` optional until we can move to c++23 (so, when CUDA supports it... plus at least few months to adopt those releases).

**Comment by bernhardmgruber** [comment]
> > I think alpaka's `ViewSubView` should be entirely replaced by `std::mdspan`
> 
> No. This would break user code that cannot depend on `std::mdspan`.

This GH issue here is about extending `ViewSubView` to support things that `std::mdspan` probably does better. So I am claiming to not implement those on `ViewSubView`, and rather use `std::mdspan` instead for this functionality. `ViewSubView` should be dropped in the long term. Even though we break API in alpaka every now and then, if you are concerned, we can give it a release or two of  `[[deprecated]]`. But `std::mdspan` should replace it in the long term, and we should not spend resources on `ViewSubView`.

> Edit: At least not until `std::mdspan` becomes available in the standard libraries.

It's available in MSVC and clang-17. I think not yet in libstdc++. CUDA's libcu++ has it since nvcc 12.0 I think, even in C++17 mode. And the Kokkos reference implementation works well.

Nobody is forcing anybody to use it. I know some users have reservations against Kokkos. But the feature will become available down the road and we should not invest in alternatives.
